staging/src/k8s.io/component-base/featuregate/feature_gate.go

```diff
@@ -309,51 +439,130 @@ func (f *featureGate) OverrideDefault(name Feature, override bool) error {
-// GetAll returns a copy of the map of known feature names to feature specs.
+// GetAll returns a copy of the map of known feature names to feature specs for the current emulationVersion.
 func (f *featureGate) GetAll() map[Feature]FeatureSpec {
 	retval := map[Feature]FeatureSpec{}
-	for k, v := range f.known.Load().(map[Feature]FeatureSpec) {
+	for k, v := range f.GetAllVersioned() {
+		spec := f.getCurrentVersion(v)
+		if spec.PreRelease == PreAlpha {
+			// The feature is not available at the emulation version.
+			continue
+		}
+		retval[k] = *f.getCurrentVersion(v)
+	}
+	return retval
+}
+
+// GetAllVersioned returns a copy of the map of known feature names to versioned feature specs.
+func (f *featureGate) GetAllVersioned() map[Feature]VersionedSpecs {
+	retval := map[Feature]VersionedSpecs{}
+	for k, v := range f.known.Load().(map[Feature]VersionedSpecs) {
 		retval[k] = v
 	}
 	return retval
 }
 
+func (f *featureGate) DeferErrorsToValidation(val bool) {
+	f.lock.Lock()
+	defer f.lock.Unlock()
+	f.deferErrorsToValidation = val
+}
+
+func (f *featureGate) SetEmulationVersion(emulationVersion *version.Version) error {
+	f.lock.Lock()
+	defer f.lock.Unlock()
+	klog.V(1).Infof("set feature gate emulationVersion to %s", emulationVersion.String())
+	f.emulationVersion.Store(emulationVersion)
+
+	// Copy existing state
+	enabledRaw := map[string]bool{}
+	for k, v := range f.enabledRaw.Load().(map[string]bool) {
+		enabledRaw[k] = v
+	}
+	// enabled map should be reset whenever emulationVersion is changed.
+	enabled := map[Feature]bool{}
+	errs := f.unsafeSetFromMap(enabled, enabledRaw)
+	if len(errs) == 0 {
+		// Persist changes
+		f.enabled.Store(enabled)
+	}
+	return utilerrors.NewAggregate(errs)
+}
+
+func (f *featureGate) EmulationVersion() *version.Version {
+	return f.emulationVersion.Load()
+}
+
+// FeatureSpec returns the FeatureSpec at the EmulationVersion if the key exists, an error otherwise.
+func (f *featureGate) FeatureSpec(key Feature) (FeatureSpec, error) {
+	if v, ok := f.known.Load().(map[Feature]VersionedSpecs)[key]; ok {
+		currentVersion := f.getCurrentVersion(v)
+		return *currentVersion, nil
+	}
+	return FeatureSpec{}, fmt.Errorf("feature %q is not registered in FeatureGate %q", key, f.featureGateName)
+}
+
 // Enabled returns true if the key is enabled.  If the key is not known, this call will panic.
 func (f *featureGate) Enabled(key Feature) bool {
+	// fallback to default behavior, since we don't have emulation version set
 	if v, ok := f.enabled.Load().(map[Feature]bool)[key]; ok {
 		return v
 	}
-	if v, ok := f.known.Load().(map[Feature]FeatureSpec)[key]; ok {
-		return v.Default
+	if v, ok := f.known.Load().(map[Feature]VersionedSpecs)[key]; ok {
+		return f.getCurrentVersion(v).Default
 	}
 
 	panic(fmt.Errorf("feature %q is not registered in FeatureGate %q", key, f.featureGateName))
 }
 
+func (f *featureGate) getCurrentVersion(v VersionedSpecs) *FeatureSpec {
+	return getCurrentVersion(v, f.EmulationVersion())
```

Comment:
Worth checking the feature gate here and just returning `v` if it's disabled?

Reply:
No. We don't check if a specific feature gate is enabled in the base feature gate library. All the `FeatureSpec` are stored in the versioned map, there is no short cut even if the `EmulationFeature` gate is disabled. It just means the emulation version is equal to the binary version.